### ISSUE TITLE
DB Deadlock For 7.9.0 Migration

### ISSUE THREAD

**Issue by tgoodsell-tempus** [concourse/concourse#8650]
## Summary

When performing the `web` node upgrade to `7.9.0`, in a rolling fashion, a deadlock occurs for the migration with the `bigint` change.

## Steps to reproduce

1. Create a multiple web deployment running on `7.8.3`, using a PostgreSQL 12 database (`POSTGRES_12_11.R20220710.01_00`).
2. Performing a rolling upgrade to `7.9.0`, running a single web node at a time.
3. Use the `"migrate", "--migrate-to-latest-version"` command when performing the first upgrade.

## Expected results

The `migrate` tool is able to acquire the appropriate locks to perform the upgrade, even while older web nodes are still running.

My expected results are interpreted from: https://concourse-ci.org/upgrading-concourse.html#upgrading-the-web-node

## Actual results

Container logs:
`error: migration '1653924132_int_to_bigint.up.sql' failed and was rolled back: pq: deadlock detected`

DB Logs:
```
2022-12-20 22:05:10.887 UTC [793672]: [38-1] db=concourse,user=concourse_rw STATEMENT:  ALTER SEQUENCE resource_config_versions_id_seq AS bigint;

ALTER TABLE build_comments
    ALTER COLUMN build_id TYPE bigint;

ALTER TABLE successful_build_outputs
    ALTER COLUMN rerun_of TYPE bigint;

ALTER TABLE jobs
    ALTER COLUMN first_logged_build_id TYPE bigint;

ALTER TABLE containers
    ALTER COLUMN meta_build_id TYPE bigint;
```
`db=concourse,user=concourse_rw ERROR:  deadlock detected`

```
INFO 2022-12-20T22:05:11.891887Z 2022-12-20 22:05:11.890 UTC [793672]: [40-1] db=concourse,user=concourse_rw DETAIL: Process 793672 waits for AccessExclusiveLock on relation 18103 of database 16426; blocked by process 791090. Process 791090 waits for AccessShareLock on relation 18054 of database 16426; blocked by process 793672. Process 793672: ALTER SEQUENCE resource_config_versions_id_seq AS bigint; ALTER TABLE build_comments ALTER COLUMN build_id TYPE bigint; ALTER TABLE successful_build_outputs ALTER COLUMN rerun_of TYPE bigint; ALTER TABLE jobs ALTER COLUMN first_logged_build_id TYPE bigint; ALTER TABLE containers ALTER COLUMN meta_build_id TYPE bigint; Process 791090: SELECT p.id, p.name, p.groups, p.var_sources, p.display, p.nonce, p.version, p.team_id, t.name, p.paused, p.public, p.archived, p.last_updated, p.parent_job_id, p.parent_build_id, p.instance_vars, p.paused_by, p.paused_at FROM pipelines p LEFT JOIN teams t ON p.team_id = t.id LEFT JOIN jobs j ON (j.id = p.parent_job_id) LEFT JOIN pipelines parent ON (j.pipeline_id = parent.id) WHERE (p.parent_job_id IS NOT NULL AND p.archived = $1 AND (j.active = $2 OR j.id IS NULL OR parent.archived = $3 OR (p.parent_build_id < j.latest_completed_build_id AND EXISTS ( SELECT 1 FROM builds lb WHERE lb.id = j.latest_completed_build_id AND lb.status = $4 ))))
```

## Additional context

The blocking process is:
```
2022-12-20 22:02:03.201 UTC [791090]: [4-1] db=concourse,user=concourse_rw STATEMENT:  DELETE FROM resource_caches WHERE id NOT IN (SELECT resource_cache_id FROM resource_cache_uses UNION SELECT resource_cache_id FROM resource_configs WHERE resource_cache_id IS NOT NULL UNION SELECT resource_cache_id FROM build_image_resource_caches UNION SELECT r_cache.id FROM next_build_inputs nbi JOIN resources r ON r.id = nbi.resource_id JOIN resource_config_versions rcv ON rcv.version_md5 = nbi.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id JOIN resource_caches r_cache ON r_cache.resource_config_id = r.resource_config_id AND r_cache.version_md5 = rcv.version_md5 JOIN jobs j ON nbi.job_id = j.id JOIN pipelines p ON j.pipeline_id = p.id WHERE p.paused = false) RETURNING id
```

## Triaging info

* Concourse version: 7.9.0
* Browser (if applicable): N/A
* Did this used to work? First time for this specific migration

**Comment by evanchaoli** [comment]
@tgoodsell-tempus When there some data changes, rolling upgrade may lead to db deadlock. Consider that, one process is changing the db schema, other processes are still operating data in the db, which is imaginable deadlock.

So, anytime when you upgrade to a new version, it's better to reduce ATC count to 0, then restart ATC with a new version, so that, only one ATC will acquire the db migration lock, other ATCs will wait till db migration finishes.

In your case, you may retry the upgrade again with the procedure:

* Reduce ATC count to 0 (shutdown all ATCs)
* Restart ATCs with 7.9.0.

If you don't urgently need some features of 7.9.0, I would suggest wait for 7.9.1, because 7.9.1 will include an important bug fix https://github.com/concourse/concourse/pull/8486.

**Comment by tgoodsell-tempus** [comment]
@evanchaoli Thanks for that information, yes that does make more sense in practice, so I can adjust our rollout strategies accordingly. Appreciate the notice about waiting till `7.9.1` as well.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
